igvm/drbd.py

```diff
@@ -0,0 +1,276 @@
+            '    }}\n'
+            '{src_host}\n'
+            '{dst_host}\n'
+            '}}\n'.format(
+                dev=self.vm_name,
+                src_host=self.get_host_config(),
+                dst_host=peer.get_host_config(),
+            )
+        )
+        self.hv.put('/etc/drbd.d/{}.res'.format(self.vm_name), fd, '0640')
+
+    def get_host_config(self):
+        return (
+            '    on {host} {{\n'
+            '        address   {addr}:{port};\n'
+            '        device    /dev/drbd{dm_minor};\n'
+            '        disk      /dev/mapper/{device}{role};\n'
+            '        meta-disk /dev/mapper/{device}_meta;\n'
+            '    }}'.format(
+                host=self.hv.dataset_obj['hostname'],
+                addr=self.hv.dataset_obj['intern_ip'],
+                port=self.get_device_port(),
+                dm_minor=self.get_device_minor(),
+                device=self.get_mapper_name(),
+                role='_orig' if self.master_role else ''
+            )
+        )
+
+    def replicate_to_slave(self, tx=None):
+        # Size must be retrieved before suspending device
+        dev_size = self.get_device_size()
+
+        # Suspend all traffic to disk from VM
+        self.hv.run('dmsetup suspend {}'.format(self.get_mapper_name()))
+        if tx:
+            tx.on_rollback(
+                'Resume original device', self.hv.run,
+                'dmsetup resume {}'.format(self.get_mapper_name())
+            )
+            # The "up" command might fail due to misconfiguration but the
+            # device is started nevertheless. This is why "down" rollback is
+            # always performed.
+            tx.on_rollback(
+                'Bring DRBD device down', self.hv.run,
+                'drbdadm down {}'.format(self.vm_name)
+            )
+
+        # Start DRBD on device
+        self.hv.run('drbdadm create-md {}'.format(self.vm_name))
+        self.hv.run('drbdadm up {}'.format(self.vm_name))
+
+        # Enforce primary operation and sync to secondary with
+        # overwriting of data
+        self.hv.run(
+            'drbdadm -- --overwrite-data-of-peer primary {}'
+            .format(self.vm_name)
+        )
+
+        # DRBD is finally up, now replace device which VM talks to on-fly.
+        # In Device Mapper block is always 512 bytes.
+        self.hv.run(
+            'dmsetup load {} --table "0 {} linear /dev/drbd{} 0"'
+            .format(
+                self.get_mapper_name(),
+                dev_size / 512,
+                self.get_device_minor(),
+            )
+        )
+        if tx:
+            tx.on_rollback(
+                # There should be no need for resume because it happens also
+                # via another rollback defined above. Unfortunately it is
+                # needed before DRBD can be shutdown.
+                # WARNING: Potental race between writes to DRBD and underlying
+                # device - potential data loss?
```

Comment:
I don't quite get the reason yet. What happens if you don't resume it before?

Reply:
I updated the comment to be more clear about this issue.